src/main/java/com/google/edith/services/SearchServiceImpl.java

```diff
@@ -0,0 +1,152 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.edith.services;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.CompositeFilter;
+import com.google.appengine.api.datastore.Query.CompositeFilterOperator;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.users.UserService;
+import com.google.common.collect.ImmutableList;
+import com.google.edith.interfaces.SearchService;
+import com.google.edith.servlets.Item;
+import com.google.edith.servlets.Receipt;
+import java.util.ArrayList;
+import java.util.List;
+
+/** Service to create query and handle searching on entities. */
+public final class SearchServiceImpl implements SearchService {
+
+  private final DatastoreService datastore;
+  private final UserService userService;
+
+  public SearchServiceImpl(DatastoreService datastore, UserService userService) {
+    this.datastore = datastore;
+    this.userService = userService;
+  }
+
+  @Override
+  public ImmutableList<Receipt> createReceiptObjects(ImmutableList<Entity> entities) {
+    List<Receipt> receipts = new ArrayList<>();
+
+    for (Entity entity : entities) {
+      Key entityKey = entity.getKey();
+      Query itemQuery = new Query("Item", entityKey);
+      PreparedQuery results = datastore.prepare(itemQuery);
+      List<Entity> itemEntities = results.asList(FetchOptions.Builder.withLimit(Integer.MAX_VALUE));
+      ImmutableList<Item> items = createItemObjects(ImmutableList.copyOf(itemEntities));
+      Receipt receipt =
+          new Receipt(
+              (String) entity.getProperty("userId"),
+              (String) entity.getProperty("storeName"),
+              (String) entity.getProperty("date"),
+              (String) entity.getProperty("name"),
+              (String) entity.getProperty("fileUrl"),
+              // Datastore returns as double even when uploaded as float.
+              (float) ((double) entity.getProperty("price")),
+              items.toArray(new Item[items.size()]));
+
+      receipts.add(receipt);
+    }
+    return ImmutableList.copyOf(receipts);
+  }
+
+  @Override
+  public ImmutableList<Item> createItemObjects(ImmutableList<Entity> entities) {
```

Comment:
This method can be static, and probably private and called from another method.

Reply:
I am calling it from a different method. The query might be to retrieve Items only so we might have to expose this method to servlet in that case.